addons/hr_expense/controllers/webmanifest.py

```diff
@@ -0,0 +1,45 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.tools import file_open
+from odoo.addons.web.controllers import webmanifest
+
+
+class WebManifest(webmanifest.WebManifest):
+
+    def _get_service_worker_content(self):
+        body = super()._get_service_worker_content()
+        with file_open('hr_expense/static/src/service_worker.js', 'r', filter_ext=('.js',)) as f:
+            content = f.read()
+            # find the begin of the body event listener fetch
+            search_body_str = 'self.addEventListener("fetch", (event) => {\n'
+            len_end_fetch = len('});\n')
+            index_fetch = body.find(search_body_str) + len(search_body_str)
+            # extract the fetch to append it inside the existing event fetch
+            begin_fetch = "// begin fetch"
+            end_fetch = "// end fetch"
+            index_begin = content.find(begin_fetch) + len(begin_fetch) + len(search_body_str)
+            index_end = content.find(end_fetch) - len_end_fetch
+            fetch_method_content = content[index_begin:index_end]
+            body = body[:index_fetch] + fetch_method_content + body[index_fetch:]
+            # append the end of the file
+            body += content[index_end + len_end_fetch + len(end_fetch):]
+        return body
```

Comment:
Gloves off, this is a very bad piece of code. Don't do such string manipulations. Not only is it hard to read but also not future-proof. What if someone changes the original body do `self.addEventlListener("fetch", (evt) => {\n`?

This must be fully rewritten.

I would very much prefer that we do not call `super()` if we are not using super's result in full. Like I would like to suggest to store a modified copy of super()'s file inside hr_expense.

If this is not possible / desirable, that string manipulation is the only way, then please make the code much more explicit. I see that you open a file to copy a function from that file and inject it in super()'s file. To try to avoid duplicating code at all cost is a bad habit, it is very fine to have the function hardcoded as a string inside python, and maybe add a comment next to the two bit of code "keep me synchronized with ...".

Please learn about [`str.partition`](https://docs.python.org/3/library/stdtypes.html#str.partition), it combines str.find and substrings, it should ease the code a lot.

```py
body = super()._get_service_worker_content()

# isolate the eventListener from the body
before, _event_listener_open, rest = body.partition('self.addEventListener("fetch", (event) => {\n')
if not _event_listener_open:
    raise ValueError("search string not found in body")
_event_listener_inner, _event_listener_close, after = rest.partition('});\n')
if not _event_listener_close:
    raise ValueError("search string not found in body")
# body == before + (_event_listener_open + _event_listener_inner + _event_listener_close) + after

# replace the eventListener by ours
# keep this code snippet synchronized with addons/hr_expense/static/src/service_worker.js
hr_expense_listener = R"""\
self.addEventListener("fetch", (event) => {
    const { pathname, searchParams } = new URL(event.request.url);
    if (
        pathname === "/odoo" &&
        event.request.method === "POST" &&
        searchParams.has("share_target")
    ) {
        serveShareTarget(event);
        return;
    }
});
"""
return before + hr_event_listener + after
```

Bonus point if you change both partition's search string to include the expected indentation. e.g. `res.partition('\n    });\n')` if we expect an indentation of 4 spaces.

Reply:
You are right, I will move the service worker code from hr_expense to the web service worker. This will be better for the future if we expand the functional scope of the share target.